### PR TITLE
Bandwidth scenario

### PR DESCRIPTION
**Pull request by maiqueb**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,31 @@
...
+providing traffic shaping between the interfaces using it.
+
+This demo scenario is composed of two net namespaces, interconnected by a linux
+bridge.
+
+We will use two different CNI configurations, to show how the bandwidth CNI
+complements the functionality of the `bridge-cni` plugin with traffic shaping.
+These configurations are:
+- [unlimitedbandwidth](examples/bandwidth/10-unlimited.conflist): no traffic shaping
+- [bandwidthlimiter](examples/bandwidth/10-limited.conflist): sets up an ingress rate & burst limiter
```
Comment (maiqueb) [inline README.md:34]: I should refer somewhere the importance of `conflist` vs `conf` file extensions ... It bit me.

File: demo-scenarios/bandwidth-example.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+if [ "$#" -ne 1 ]; then
+    echo "$0 <name of the network configuration to use>"
+    exit 1
+fi
+
+bandwidth_net_name="$1"
+
+export NETCONFPATH="${NETCONFPATH:-(pwd)/examples}"
+export CNI_PATH="${CNI_PATH:-../cni-plugins/bin}"
```
Comment (dougbtv) [inline demo-scenarios/bandwidth-example.sh]: Went to run it and this path doesn't exist. Should there be an install section about building the CNI plugins? in my testing/review I'm going to change it to path where I have all the CNI plugins built.

Comment (dougbtv) [reply demo-scenarios/bandwidth-example.sh]: (or you could include the bins in the repo, I guess, for demo purposes)

File: demo-scenarios/bandwidth-example.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+# start the server "pod"
+sudo ip netns add server
+sudo NETCONFPATH=$(echo $NETCONFPATH) \
+    CNI_PATH=$(echo $CNI_PATH) \
+    CNI_ARGS="IgnoreUnknown=true;IP=$serverIP/24" \
+    CNI_IFNAME=ens123 \
+    CNI_CONTAINERID=$(uuidgen) \
+    $(which cnitool) add $bandwidth_net_name "/var/run/netns/server"
+sudo ip netns exec server bash -c "iperf3 -s -p 9000 -1 &"
```
Comment (dougbtv) [inline demo-scenarios/bandwidth-example.sh:28]: I didn't have iperf3 on my (fedora) workstation, I just installed with `dnf install iperf3` fwiw.

Comment (dougbtv) [reply demo-scenarios/bandwidth-example.sh:28]: Additionally, I also got...

```
+ sudo ip netns exec client bash -c 'iperf3 -c 192.168.200.3 -p 9000 -t=5'
iperf3: error - unable to connect to server: Connection refused
```

**Comment by dougbtv** [comment]
Ran into a couple things on my own run-through of it.

Unsure how bootstrapped you want it to be, or if you want your audience to be able to pick it up and use it immediately, but...

...that's what I found.

Don't get me wrong though -- I think you're on the right path. I also think that bandwidth plugin is a VERY intuitive and good demo for CNI plugin chains. I think it's a really good example of how there's these kind of "meta things" that you want to do that aren't explicitly creating network interfaces but manipulation of them that are perfect for showing how CNI chains work and the the power of the modularity.

File: demo-scenarios/bandwidth-example.sh (hunk 3 above)
Comment (dougbtv) [reply demo-scenarios/bandwidth-example.sh:28]: full results @ https://pastebin.com/hSwYKE83 (expires 6 mos)

File: demo-scenarios/bandwidth-example.sh (hunk 2 above)
Comment (danielmellado) [reply demo-scenarios/bandwidth-example.sh]: I'd put the bins in the repo directly, either that or just the default one /opt/cni/bin?

Comment (maiqueb) [reply demo-scenarios/bandwidth-example.sh]: The user can point the script to where ever she/he wants by using this variable though. 

I'd rather not ship any binaries here, not use submodules / build them.

I'd rather add a `requirements` heading to the `README` indicating what steps must be followed as pre-requirements (build the plugins from CNI, install iperf, etc).

Thoughts ?

Comment (maiqueb) [reply demo-scenarios/bandwidth-example.sh]: > I'd put the bins in the repo directly, either that or just the default one /opt/cni/bin?

I agree this default is more appropriate than the one currently provided. Either that, or explode spectacularly if the user does not provide this path.

**Comment by danielmellado** [comment]
Hi folks, I thought that `git-pull-request`would allow me to have a pull request to Miguel's PR but it created a new one, https://github.com/maiqueb/fosdem2023-cni-unchained/pull/2! xD.

TL;DR Sanitized the bash variables and exports and bundled both cni plugin and cnitool binaries. I think it works in an easier way now. PTAL!

**Comment by maiqueb** [comment]
Hey @danielmellado ; I've cherry-picked your commit from #2 into this one, and added a tiny script that downloads and builds cnitool / ref plugins from src.

Thoughts ?
